src/System.Management.Automation/engine/ScriptCommandProcessor.cs

```diff
@@ -633,5 +627,143 @@ protected override void OnRestorePreviousScope()
                 CommandSessionState.CurrentScope.DottedScopes.Pop();
             }
         }
+
+        private bool _cleanupCompleted;
+
+        internal void InvokeCleanupBlock()
+        {
+            if (_cleanupCompleted)
+            {
+                return;
+            }
+
+            if (!_scriptBlock.HasCleanupBlock)
+            {
+                ScriptBlock.LogScriptBlockEnd(_scriptBlock, Context.CurrentRunspace.InstanceId);
+                return;
+            }
+
+            ExecutionContext currentContext = LocalPipeline.GetExecutionContextFromTLS();
+            if (Context != currentContext)
+            {
+                // Something went wrong; Cleanup {} is being called from the wrong thread.
+                // Create an event to call it from the correct thread.
+                InvokeCleanupInOriginalContext();
+                return;
+            }
+
+            Pipe oldOutputPipe = _functionContext?._outputPipe;
+            try
+            {
+                Action<FunctionContext> cleanupBlock = _runOptimizedCode
+                    ? _scriptBlock.CleanupBlock
+                    : _scriptBlock.UnoptimizedCleanupBlock;
+
+                if (_functionContext != null)
+                {
+                    _functionContext._outputPipe = new Pipe
+                    {
+                        NullPipe = true
+                    };
+                }
+
+                // run with no pipeline input or $input enumerator
+                RunClause(cleanupBlock, AutomationNull.Value, AutomationNull.Value);
+            }
+            finally
+            {
+                if (_functionContext != null)
+                {
+                    _functionContext._outputPipe = oldOutputPipe;
+                }
+
+                ScriptBlock.LogScriptBlockEnd(_scriptBlock, Context.CurrentRunspace.InstanceId);
+
+                _cleanupCompleted = true;
```

Comment:
In theory, I think it's possible. Dunno how likely it would be in practice, though.

We could lock this on a per-ScriptCommandProcessor-basis to break that, I think, yeah. That might be a good idea. 🤔

Reply:
Should we also then have a lock around the `Dispose()` method for `CompiledScriptBlock` to mirror this, then?